raft: Prevent the same manager from joining raft twice

aaronlehmann opened the pull request:
If a manager's raft state is lost or corrupted, it may try to join the
Raft cluster as a fresh member. This must be avoided, because it will
influence the quorum of the Raft cluster, which can prevent leader
elections from succeeding.

Add a check to `Join` that prevents a node with the same identity from
joining twice.

cc @cyli @LK4D4

cyli commented on manager/state/raft/raft.go:
```diff
@@ -710,11 +710,20 @@ func (n *Node) Join(ctx context.Context, req *api.JoinRequest) (*api.JoinRespons
+	// A single manager must not be able to join the raft cluster twice. If
+	// it did, that would cause the quorum to be computed incorrectly. This
+	// could happen if the WAL was deleted from an active manager.
+	for _, m := range n.cluster.Members() {
```
In the case of the raft on-disk logs being deleted from all the active manager, before it rejoins the cluster, would the `n.cluster` object have any members?

cyli replied:
Hmm... nm I think I misunderstood the comment `This method is called from an aspiring member to an existing member`

aaronlehmann commented:
I've added another commit with a related safeguard: `ProcessRaftMessage` should only process messages that were intended for this node's `RaftID`. I believe the situation that arises in  https://github.com/docker/docker/issues/29180#issuecomment-265518804 involves the same node joining raft a second time, and then receiving messages intended for the original raft member. It turns out there was no check filtering such messages out, so they could trigger a panic.

The first commit with the changes to `Join` should protect against this, but it seems like a good idea to also be defensive when receiving messages, for example in case the leader is running an old version that allows `Join` to be called twice.

I've also added a test that triggers the panic on master.

Let me be clear about which cases leading to the panic this PR does and doesn't fix:

1. Wiped state. This is the only case of the three that is fixed. Previously, wiping the state would cause the node to join raft under a second ID, which messed up the quorum calculation and could cause the panic when the node receives messages intended for the old raft ID. This PR prevents that. 
2. Failed writes (for example, full disk). It turns out we're calling `Advance` even if there was an error writing to disk. This is bad because it tells etcd/raft that data was committed successfully, even if that is not true. It can lead to a later panic when etcd/raft discovers it's missing data that it thought it had. However, it's not obvious what to do here instead. We could panic right away when the write fails (this is what etcd does), or return an error from `Run` so the manager shuts down (which is still disruptive, but slightly less so, and may lead to deadlocks, because `Run` isn't really supposed to return spontaneously).
3. Downgrades. Because of the data migration, a downgraded node can "go back in time". We thought this would work, but it violates assumptions in etcd/raft. We probably need to remove the old WAL and snapshot directories (unfortunate, since we explicitly kept these around to allow downgrades), and also document a downgrade process that involves removing the upgraded manager and adding a new one on the older version.

cc @dongluochen

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -990,6 +999,11 @@ func (n *Node) ProcessRaftMessage(ctx context.Context, msg *api.ProcessRaftMessa
 
 	if n.IsMember() {
+		if msg.Message.To != n.Config.ID {
+			n.processRaftMessageLogger(ctx, msg).Errorf("received message indented for raft_id %x", msg.Message.To)
```
I think it should be intended.

aaronlehmann replied:
Fixed :)

LK4D4 commented:
LGTM
Feel free to merge after misspell fix.
